packages/back-end/src/util/organization.util.ts

```diff
@@ -68,10 +71,77 @@ export function roleToPermissionMap(
   return permissionsObj;
 }
 
-export function getUserPermissions(
+function mergePermissions(
+  existingPermissions: UserPermission,
+  existingRole: MemberRole | undefined,
+  teamInfo: TeamInterface | ProjectMemberRole,
+  org: OrganizationInterface
+) {
+  const newPermissions = roleToPermissionMap(teamInfo.role, org);
+  for (const newPermission in newPermissions) {
+    // If the user doesn't have permission, but the team role does, add it
+    if (
+      !existingPermissions.permissions[newPermission as Permission] &&
+      newPermissions[newPermission as Permission]
+    ) {
+      existingPermissions.permissions[newPermission as Permission] =
+        newPermissions[newPermission as Permission];
+    }
+  }
+
+  if (!existingRole) {
+    existingPermissions.limitAccessByEnvironment =
+      teamInfo.limitAccessByEnvironment;
+    existingPermissions.environments = teamInfo.environments;
+  } else {
+    const roles = getRoles(org);
+    const newRoleIndex = roles.findIndex((r) => r.id === teamInfo.role);
+    const existingRoleIndex = roles.findIndex((r) => r.id === existingRole);
+
+    const existingRoleIsEngineerOrExperimenter =
+      existingRole === "engineer" || existingRole === "experimenter";
+
+    const newRoleIsEngineerOrExperimenter =
+      teamInfo.role === "engineer" || teamInfo.role === "experimenter";
+
+    if (
+      existingRoleIsEngineerOrExperimenter &&
+      newRoleIsEngineerOrExperimenter
+    ) {
+      if (
+        // If limitAccessByEnvironment is the same for new and existing roles, we just concat the envs arrays
+        existingPermissions.limitAccessByEnvironment ===
+        teamInfo.limitAccessByEnvironment
+      ) {
+        existingPermissions.environments = [
+          ...new Set(
+            existingPermissions.environments.concat(teamInfo.environments)
+          ),
+        ];
+      } else {
+        // otherwise, 1 role doesn't have limited access by environment, so it overrides the other
+        existingPermissions.limitAccessByEnvironment = false;
+        existingPermissions.environments = [];
+      }
+    } else {
+      // Finally, we set the limitAccessByEnvironment and environments properties to the more permissive role's values.
+      existingPermissions.limitAccessByEnvironment =
+        newRoleIndex > existingRoleIndex
+          ? teamInfo.limitAccessByEnvironment
+          : existingPermissions.limitAccessByEnvironment;
+
+      existingPermissions.environments =
+        newRoleIndex > existingRoleIndex
+          ? teamInfo.environments
+          : existingPermissions.environments;
+    }
+  }
+}
+
+export async function getUserPermissions(
   userId: string,
   org: OrganizationInterface
-): UserPermissions {
+): Promise<UserPermissions> {
   const memberInfo = org.members.find((m) => m.id === userId);
   const userPermissions: UserPermissions = {
```

Comment:
Should we early return this `UserPermissions` object if we don't find the `memberInfo` or throw an error?

Reply:
Great suggestion - that cleans up the code quite a bit - I can get rid of so many `?`'s lol. I also added a test for this use case.